### PR TITLE
netcat: explicit use of IPv4 and IPv6

### LINKED ISSUE #3213: `./netcat -l --udp localhost 8080` fail to write connection - destination address required
**Describe the bug**

Following error were observed from running netcat command to start listening on 8080, and then there is an error complaining "failed to write to connection: ..."

```
xuehaohu@worldpeace:~/go/src/github.com/u-root/u-root/cmds/core/netcat$ ./netcat -l --udp localhost 8080

2024/11/30 08:53:41 failed to write to connection: write udp 127.0.0.1:8080: write: destination address required

```

(**From another shell**)

Trying to connect would give this error,

```
xuehaohu@worldpeace:~/go/src/github.com/u-root/u-root/cmds/core/netcat$ ./netcat -C --udp localhost 8080
2024/11/30 09:22:05 Connection to localhost:8080 [udp] succeeded

2024/11/30 09:22:07 error: failed to write: read udp [::1]:40723->[::1]:8080: read: connection refused
xuehaohu@worldpeace:~/go/src/github.com/u-root/u-root/cmds/core/netcat$ ./netcat -C --udp localhost 8080
2024/11/30 09:22:14 Connection to localhost:8080 [udp] succeeded

2024/11/30 09:22:15 error: failed to write: read udp [::1]:42713->[::1]:8080: read: connection refused
xuehaohu@worldpeace:~/go/src/github.com/u-root/u-root/cmds/core/netcat$
```

**To Reproduce**

` ./netcat -l --udp localhost 8080`

**Expected behavior**

no confusing error. 

**Additional context**

### PR DESCRIPTION
**Pull request by jensdrenhaus**
Fixes #3213

### CODE REVIEW

**Comment by jensdrenhaus** [comment]
Still need to do final testing and improve some logging.

File: cmds/core/netcat/connect.go [hunk 1]
```diff
@@ -28,6 +28,24 @@ import (
 var osConnectors = map[netcat.SocketType]func(string, string) (net.Conn, error){}
 
 func (c *cmd) connectMode(output io.Writer, network, address string) error {
+	if network == "tcp" || network == "udp" {
+		network4 := network + "4"
+		err4 := c.connect(output, network4, address)
```
Comment (rminnich) [inline cmds/core/netcat/connect.go:33]: or you can just
err4 := c.connect(output, network+"4", address)

File: cmds/core/netcat/listen.go [hunk 2]
```diff
@@ -24,6 +24,24 @@ import (
...
+		err4 := c.listen(output, network4, address)
+		if err4 == nil {
+			return nil
+		}
+		network6 := network + "6"
+		err6 := c.listen(output, network6, address)
+		if err6 == nil {
+			return nil
+		}
+		return fmt.Errorf("listen IPv4: %w\nlisten IPv6: %wgi", err4, err6)
```
Comment (rminnich) [inline cmds/core/netcat/listen.go:38]: I don't think you should embed newlines in an error message. You could use a single error variable and errors.Join I think.

Comment (rminnich) [reply cmds/core/netcat/listen.go:38]: also, does the err have information such that you don't need the "listen IPV4"?
I suspect so.
You might just need fmt.Errorf("listenMode:%w", and you could have a single err and  errors.Join

**Review by rminnich — CHANGES_REQUESTED** [review]

File: cmds/core/netcat/listen.go (hunk 2 above)
Comment (jensdrenhaus) [reply cmds/core/netcat/listen.go:38]: Nice, errors.Join seem decent!

**Comment by jensdrenhaus** [comment]
BTW, any idea why we do not allow squash and merge? So that I could add commits addressing comments and such without the need to force push and thus messing up the PR conversations history.

**Review by binjip978 — DISMISSED** [review]

File: cmds/core/netcat/connect_test.go [hunk 3]
```diff
@@ -20,6 +22,113 @@ import (
 	"github.com/u-root/u-root/pkg/netcat"
 )
 
+func TestConnectMode(t *testing.T) {
+	tests := []struct {
+		name        string
+		network     string
+		address     string
+		connectFunc func(output io.Writer, network, address string) error
+		expectError bool
```
Comment (leongross) [inline cmds/core/netcat/connect_test.go:31]: If we have test cases that throw known errors, wouldn't it make more sense to embed them into the test cases? So either it is `nil` or the error we predefined?

 ```suggestion
	     expectedError error
```

and then we can check for the result
```go
if err := cmd.connectMode(io.Discard, tt.network, tt.address); err != tt.expectedError {
    t.Errorf("wanted %v, got %v", tt.expectedError, err)
}
```

File: cmds/core/netcat/listen_test.go [hunk 4]
```diff
@@ -18,6 +18,113 @@ import (
 	"github.com/u-root/u-root/pkg/netcat"
 )
 
+func TestListenMode(t *testing.T) {
+	tests := []struct {
+		name        string
+		network     string
+		address     string
+		listenFunc  func(output io.Writer, network, address string) error
+		expectError bool
```
Comment (leongross) [inline cmds/core/netcat/listen_test.go:27]: Same here
```suggestion
		expectedError error
```

File: cmds/core/netcat/connect_test.go (hunk 3 above)
Comment (jensdrenhaus) [reply cmds/core/netcat/connect_test.go:31]: If there are defined error values, yes, sure. This would make sense in the actual [connect func](https://github.com/u-root/u-root/pull/3291/files/2c13f0513578ad8168ab7f2c0a4ba427f78c643f#diff-1445756be1283b81274f42febc4561c59c56b8d236f6b49317e23eb5d394c3edR46) which is the former 'ConnectMode()' function. 

What I did here is pushing the code of the former 'ConnectMode()' function into 'connect()' and added logic in the new 'ConnectMode()' function to call 'connect()' several times in some cases. To test this logic I'm using a stub function instead of 'connect()' that just return errors as needed. I think this is sufficient to test the introduced logic. 

Adding defined error types and testing against them would be in the scope of [TestConnect](https://github.com/u-root/u-root/pull/3291/files/2c13f0513578ad8168ab7f2c0a4ba427f78c643f#diff-16aa8d474e6f1d06d5edc08b989c613ff40f2854257292332a816e3f325a8a9eR152) and is out of scope for this PR.

File: cmds/core/netcat/listen_test.go (hunk 4 above)
Comment (jensdrenhaus) [reply cmds/core/netcat/listen_test.go:27]: See comment on [TestConnectMode](https://github.com/u-root/u-root/pull/3291/files/2c13f0513578ad8168ab7f2c0a4ba427f78c643f#diff-16aa8d474e6f1d06d5edc08b989c613ff40f2854257292332a816e3f325a8a9eR25)

File: cmds/core/netcat/connect_test.go (hunk 3 above)
Comment (leongross) [reply cmds/core/netcat/connect_test.go:31]: I understand. Let's keep it the way it is.

**Review by leongross — DISMISSED** [review]

**Comment by jensdrenhaus** [comment]
@rminnich ready for review. Your change request should be addressed now.

File: cmds/core/netcat/connect_test.go [hunk 5]
```diff
@@ -20,6 +22,110 @@ import (
...
 
+func TestConnectMode(t *testing.T) {
+	connectErr := errors.New("failed to connect")
+
+	tests := []struct {
+		name          string
+		network       string
+		address       string
+		connectFunc   func(output io.Writer, network, address string) error
+		expectedError error
```
Comment (rminnich) [inline cmds/core/netcat/connect_test.go:33]: one little note here: go style tends to be brief, it's more comment to name this err and not expectedError.
See what you think. I personally prefer just err.

File: cmds/core/netcat/connect_test.go [hunk 6]
```diff
@@ -20,6 +22,110 @@ import (
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cmd := &cmd{
+				connectStubFn: tt.connectFunc,
+			}
+			err := cmd.connectMode(io.Discard, tt.network, tt.address)
+			if !errors.Is(err, tt.expectedError) {
+				t.Errorf("Expected error %v, got %v", tt.expectedError, err)
```
Comment (rminnich) [inline cmds/core/netcat/connect_test.go:122]: The common pattern is the 'got, want' pattern, I believe, i.e. what you got, and what you expected. 
Certainly in my code reviews at Google that's what they'd tell me. 

It makes it easier to scan the logs in Ci if we keep this consistent. again, I leave it to you.

File: cmds/core/netcat/netcat.go [hunk 7]
```diff
@@ -335,6 +335,10 @@ type cmd struct {
 	stderr io.Writer
 	config *netcat.Config
 	args   []string
+
```
Comment (rminnich) [inline cmds/core/netcat/netcat.go:338]: This is now how it should be done, sorry. 

The existence of these indicated a problem with the design. Let's reconsider this.

**Review by rminnich — CHANGES_REQUESTED** [review]
connectStubFn and listenStubFn are not a good way to do this, let's slow down a bit and get it right.

File: cmds/core/netcat/connect_test.go (hunk 6 above)
Comment (jensdrenhaus) [reply cmds/core/netcat/connect_test.go]: Agree, consistency matters.

File: cmds/core/netcat/netcat.go (hunk 7 above)
Comment (jensdrenhaus) [reply cmds/core/netcat/netcat.go:338]: Yeah, I was wondering if this was a wise approach or not. Any alternative suggestions would be appreciated. 

The thing is that the _listen_ and _connect_ functions of netcat are basically wrappers around some net package functionality (or tls package, et al, that work with `net.Conn` and `net.Listener`). I think we should not test the functionality of the net package here, but rather isolate the logic added by netcat and test that remaining part. 

I have not delved into refactoring the existing connect and listen functions and their testing. Pretty sure there is potential for improvement in the above manner. 
To me, the way I added the logic of the change seems somewhat decent, and I was looking for a way to add meaningful testing for that added logic. 

Again, glad to see an alternative approach.

**Comment by rminnich** [comment]
create a mockCmd struct that wraps command
implement connect for that struct
something like
struct mockCmd{
cmd
network string
}

func(m*mockCmd) connect(...)error{
if m.network == network
return nil
return connectErr
}
or some such

do the same for the other function

see mocks at end of the test file as an example

**Comment by jensdrenhaus** [comment]
> create a mockCmd struct that wraps command implement connect for that struct something like struct mockCmd{ cmd network string }
> 
> func(m*mockCmd) connect(...)error{ if m.network == network return nil return connectErr } or some such
> 
> do the same for the other function
> 
> see mocks at end of the test file as an example

Thanks, @rminnich for that input, always fascinated by the power of embedding in Go. 
Unfortunately it does not work out here. 

I want to test cmd.connectMode() function. This function is calling cmd.connect(), so I need to mock/inject cmd.connect(). 

If I use your approach like so: 

```
type mockCmd struct {
	cmd
	connectStubFn func(output io.Writer, network, address string) error
}

func (m *mockCmd) connect(output io.Writer, network, address string) error {
	if m.connectStubFn == nil {
		return errors.New("mockCmd: connectStubFn is not set")
	}
	return m.connectStubFn(output, network, address)
}

func TestConnectMode(t *testing.T) {
	connectErr := errors.New("failed to connect")

	tests := []struct {
		name        string
		network     string
		address     string
		connectFunc func(output io.Writer, network, address string) error
		err         error
	}{
		{
			name:    "TCPv4 success",
			network: "tcp",
			address: "localhost:8080",
			connectFunc: func(output io.Writer, network, address string) error {
				if network == "tcp4" {
					return nil
				}
				return connectErr
			},
			err: nil,
		},

// ...

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cmd := &mockCmd{
				connectStubFn: tt.connectFunc,
			}
			err := cmd.connectMode(io.Discard, tt.network, tt.address)
			if !errors.Is(err, tt.err) {
				t.Errorf("got %v, want %v", err, tt.err)
			}
		})
	}

}
```
mockCmd.connect will not be called but cmd.connect(). 

Pls correct me if I misunderstood. 

What about classic DI like: 
```
type connectFn func(io.Writer, string, string) error

func (c *cmd) connectMode(output io.Writer, network, address string, fn connectFn) error {
...
}
```

**Comment by MDr164** [comment]
> BTW, any idea why we do not allow squash and merge? So that I could add commits addressing comments and such without the need to force push and thus messing up the PR conversations history.

It messes up the merge commit message, at least that was the reason I remember was the most crucial one. But GitHub allows you to see differences quite easily even after a force push to the branch imo.

**Review by rminnich — APPROVED** [review]
